File: Python-Projects/File-IO/student_manager.py
import os

# Function to add a student record to the file
def add_student():
    student_id = input("Enter student ID: ")
    name = input("Enter student name: ")
    grade = input("Enter student grade: ")

    with open("students.txt", "a") as file:
        file.write(f"ID: {student_id}, Name: {name}, Grade: {grade}\n")

    print("Student added successfully!")

# Function to view all students
def view_students():
    if not os.path.exists("students.txt"):
        print("No students to display.")
        return
    
    with open("students.txt", "r") as file:
        for line in file:
            print(line, end="")
    
    print()

# Function to search for a student by ID
def search_student():
    search_id = input("Enter the student ID to search: ")

    if not os.path.exists("students.txt"):
        print("No students to search.")
        return
    
    found = False
    with open("students.txt", "r") as file:
        for line in file:
            if f"ID: {search_id}" in line:
                print(f"Record found: {line}", end="")
                found = True
                break
    
    if not found:
        print(f"Student with ID {search_id} not found.")

# Function to update a student's record
def update_student():
    search_id = input("Enter the student ID to update: ")

    if not os.path.exists("students.txt"):
        print("No students to update.")
        return
    
    found = False
    with open("students.txt", "r") as file:
        lines = file.readlines()
    
    with open("temp.txt", "w") as temp_file:
        for line in lines:
            if f"ID: {search_id}" in line:
                print(f"Current Record: {line}", end="")
                name = input("Enter new name: ")
                grade = input("Enter new grade: ")
                temp_file.write(f"ID: {search_id}, Name: {name}, Grade: {grade}\n")
                found = True
            else:
                temp_file.write(line)
    
    os.remove("students.txt")
    os.rename("temp.txt", "students.txt")

    if found:
        print("Student updated successfully.")
    else:
        print(f"Student with ID {search_id} not found.")

# Function to delete a student's record
def delete_student():
    search_id = input("Enter the student ID to delete: ")

    if not os.path.exists("students.txt"):
        print("No students to delete.")
        return
    
    found = False
    with open("students.txt", "r") as file:
        lines = file.readlines()
    
    with open("temp.txt", "w") as temp_file:
        for line in lines:
            if f"ID: {search_id}" in line:
                print(f"Deleting record: {line}", end="")
                found = True
            else:
                temp_file.write(line)
    
    os.remove("students.txt")
    os.rename("temp.txt", "students.txt")

    if found:
        print("Student deleted successfully.")
    else:
        print(f"Student with ID {search_id} not found.")

# Main function to run the menu
def main():
    while True:
        print("\nStudent Management System")
        print("1. Add Student")
        print("2. View Students")
        print("3. Search Student")
        print("4. Update Student")
        print("5. Delete Student")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            add_student()
        elif choice == '2':
            view_students()
        elif choice == '3':
            search_student()
        elif choice == '4':
            update_student()
        elif choice == '5':
            delete_student()
        elif choice == '6':
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()

File: C++-Projects/File-IO/student_manager.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>

struct Student {
    int id;
    std::string name;
    int grade;
};

// Function to add a student record to the file
void addStudent() {
    Student student;
    std::cout << "Enter student ID: ";
    std::cin >> student.id;
    std::cout << "Enter student name: ";
    std::cin.ignore();  // To ignore newline left in the input buffer
    std::getline(std::cin, student.name);
    std::cout << "Enter student grade: ";
    std::cin >> student.grade;

    std::ofstream outFile("students.txt", std::ios::app); // Open file in append mode
    if (outFile.is_open()) {
        outFile << "ID: " << student.id << ", Name: " << student.name << ", Grade: " << student.grade << "\n";
        outFile.close();
        std::cout << "Student added successfully!\n";
    } else {
        std::cout << "Unable to open file.\n";
    }
}

// Function to view all students
void viewStudents() {
    std::ifstream inFile("students.txt");
    if (inFile.is_open()) {
        std::string line;
        while (getline(inFile, line)) {
            std::cout << line << "\n";
        }
        inFile.close();
    } else {
        std::cout << "Unable to open file.\n";
    }
}

// Function to search for a student by ID
void searchStudent() {
    int searchID;
    std::cout << "Enter the student ID to search: ";
    std::cin >> searchID;

    std::ifstream inFile("students.txt");
    if (inFile.is_open()) {
        std::string line;
        bool found = false;
        while (getline(inFile, line)) {
            if (line.find("ID: " + std::to_string(searchID)) != std::string::npos) {
                std::cout << "Record found: " << line << "\n";
                found = true;
                break;
            }
        }
        inFile.close();
        if (!found) {
            std::cout << "Student with ID " << searchID << " not found.\n";
        }
    } else {
        std::cout << "Unable to open file.\n";
    }
}

// Function to update a student's record
void updateStudent() {
    int searchID;
    std::cout << "Enter the student ID to update: ";
    std::cin >> searchID;

    std::ifstream inFile("students.txt");
    std::ofstream tempFile("temp.txt");
    if (inFile.is_open() && tempFile.is_open()) {
        std::string line;
        bool found = false;
        while (getline(inFile, line)) {
            if (line.find("ID: " + std::to_string(searchID)) != std::string::npos) {
                std::cout << "Current Record: " << line << "\n";
                Student student;
                std::cout << "Enter new name: ";
                std::cin.ignore();
                std::getline(std::cin, student.name);
                std::cout << "Enter new grade: ";
                std::cin >> student.grade;
                tempFile << "ID: " << searchID << ", Name: " << student.name << ", Grade: " << student.grade << "\n";
                found = true;
            } else {
                tempFile << line << "\n";
            }
        }
        inFile.close();
        tempFile.close();

        // Replace original file with updated one
        remove("students.txt");
        rename("temp.txt", "students.txt");

        if (found) {
            std::cout << "Student updated successfully.\n";
        } else {
            std::cout << "Student with ID " << searchID << " not found.\n";
        }
    } else {
        std::cout << "Unable to open file.\n";
    }
}

// Function to delete a student's record
void deleteStudent() {
    int searchID;
    std::cout << "Enter the student ID to delete: ";
    std::cin >> searchID;

    std::ifstream inFile("students.txt");
    std::ofstream tempFile("temp.txt");
    if (inFile.is_open() && tempFile.is_open()) {
        std::string line;
        bool found = false;
        while (getline(inFile, line)) {
            if (line.find("ID: " + std::to_string(searchID)) != std::string::npos) {
                std::cout << "Deleting record: " << line << "\n";
                found = true;
            } else {
                tempFile << line << "\n";
            }
        }
        inFile.close();
        tempFile.close();

        // Replace original file with updated one
        remove("students.txt");
        rename("temp.txt", "students.txt");

        if (found) {
            std::cout << "Student deleted successfully.\n";
        } else {
            std::cout << "Student with ID " << searchID << " not found.\n";
        }
    } else {
        std::cout << "Unable to open file.\n";
    }
}

int main() {
    int choice;
    do {
        std::cout << "\nStudent Management System\n";
        std::cout << "1. Add Student\n";
        std::cout << "2. View Students\n";
        std::cout << "3. Search Student\n";
        std::cout << "4. Update Student\n";
        std::cout << "5. Delete Student\n";
        std::cout << "6. Exit\n";
        std::cout << "Enter your choice: ";
        std::cin >> choice;

        switch (choice) {
            case 1:
                addStudent();
                break;
            case 2:
                viewStudents();
                break;
            case 3:
                searchStudent();
                break;
            case 4:
                updateStudent();
                break;
            case 5:
                deleteStudent();
                break;
            case 6:
                std::cout << "Exiting...\n";
                break;
            default:
                std::cout << "Invalid choice. Please try again.\n";
        }
    } while (choice != 6);

    return 0;
}

